### PR TITLE
fix(repl): support transforming JSX/TSX

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/16771

### CODE REVIEW

**Comment by dsherret** [comment]
@marvinhagemeister can you post an example where it didn't work previously? It was working for me before, but maybe it doesn't work in complex examples. I think we want to bypass swc's jsx_automatic like it was doing before because the factories will be global in the repl scope and don't want to inject an import declaration each time.

**Comment by marvinhagemeister** [comment]
@dsherret and I had a call together and we came to the conclusion that the easiest way for now is to stub out the new transform with the classic one behind the scenes. Reason being that the new transform requires some modification to work properly in the repl. The way it's done now should solve most of the common cases in the repl.

In the future we might revisit this.

**Comment by rgbkrk** [comment]
Branch looking good.

image

`@jsx` pragma working well too

image

However I couldn't use the `jsxImportSource` pragma. I'm assuming this is ok and users have to be explicit (like above).

image

File: cli/tools/repl/session.rs [hunk 1]
```diff
@@ -626,3 +733,120 @@ impl Visit for ImportCollector {
...
+    capture_tokens: true,
+    maybe_syntax: None,
+    scope_analysis: false,
+  })?;
+
+  Ok(parsed)
+}
+
+/// Matches the `@jsxImportSource` pragma.
+static JSX_IMPORT_SOURCE_RE: Lazy<Regex> =
```
Comment (dsherret) [inline cli/tools/repl/session.rs:760]: Before we merge this, we should open a PR in deno_graph to expose these and then add a todo here to remove this code.

**Review by dsherret — APPROVED** [review]
LGTM

**Comment by rgbkrk** [comment]
Woo! Delighted to see this land.

image

**Comment by rhzone** [comment]
Great work on JSX/TSX support in REPL! @bartlomieju 

However, I'm still experiencing a TypeScript language server issue with deno jupyter files: `Cannot find name 'div'. deno-ts(2304)` for JSX elements, even though the code runs perfectly.

The runtime transformation works fine, but IntelliSense doesn't recognize JSX syntax in `.ipynb` files. Is this a separate issue with the TypeScript service configuration?

image


Related issue: https://github.com/denoland/deno/issues/21079#issue-1976993964

**Comment by nayeemrmn** [comment]
@rhzone Does the same code work in a regular `.tsx` file? I think you have to run `deno add npm:@types/react`.

**Comment by rhzone** [comment]
@nayeemrmn Yes, it works perfectly in `.tsx` files. Here's my current `deno.json` configuration:

```json
{
  "tasks": {
    "dev": "deno run --watch main.ts"
  },
  "imports": {
    "@std/assert": "jsr:@std/assert@1",
    "@types/react": "npm:@types/react@^19.0.0",
    "react": "npm:react@^19.0.0"
  },
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "react",
    "jsxImportSourceTypes": "npm:@types/react@^19.0.0"
  },
  "nodeModulesDir": "auto"
}

**Pull request merged**